Сделал

vzeten opened the pull request.

Commit:
Сделал

Commit:
Сделал

7glebalyot commented on src/main/java/school/faang/user_service/controller/mentorship/MentorshipController.java:
```diff
@@ -0,0 +1,44 @@
+    }
+
+    // Посмотреть всех менти пользователя
+    @GetMapping("/mentee/{UserId}/")
```
У тебя везде в маппингах где написан {UserId}, в методах используется userId (с маленькой буквы). Лучше в маппингах изменить на вариант с маленькой буквы, так корректнее и по-моему Spring чувствителен к этому

Kalex-k replied:
Опечатка в указания пути. Лишний слэш и параметр у тебя с меленькой буквы указан, спринг не распознает его, исправь либо путь на `@GetMapping("/mentee/{userId}")` Либо явно укажи параметр через атрибут, что бы спринг его связал. `@PathVariable(name = "UserId") long userId`

7glebalyot commented on src/main/java/school/faang/user_service/controller/mentorship/MentorshipController.java:
```diff
@@ -0,0 +1,44 @@
+    }
+
+    // Посмотреть всех менторов пользователя
+    @GetMapping("/mentor/{UserId}/")
```
Так же пробегись посмотри, у тебя встречаются на конце лишние слэши "/". Не знаю на сколько это мешает коду и помешает ли в будущем, но эти слэши лишние. Ты как бы написал слэш что является ссылкой на следующую запись, но в твоём случае там пусто

Kalex-k replied:
Аналогично с getMentees

7glebalyot commented on src/main/java/school/faang/user_service/controller/mentorship/MentorshipController.java:
```diff
@@ -0,0 +1,44 @@
+    }
+
+    // Удалить связь ментор-менти
+    @DeleteMapping("/{mentorId}/delete/{menteeId}")
```
Тут также, DeleteMapping, говорит сам за себя, можно убрать delete

vzeten replied:
ок!

7glebalyot commented on src/main/java/school/faang/user_service/controller/mentorship/MentorshipController.java:
```diff
@@ -0,0 +1,44 @@
+    private final MentorshipService mentorshipService;
+
+    // Добавляем связь ментор-менти
+    @PostMapping("/{mentorId}/add/{menteeId}")
```
PostMapping уже говорит сам за себя, он добавляет "POST". В REST, когда кто-то вызывает POST, то итак понятно что он хочет добавить. Поэтому можно убрать лишний add

vzeten replied:
Точняк, спасибо!

7glebalyot commented on src/main/java/school/faang/user_service/controller/mentorship/MentorshipController.java:
```diff
@@ -0,0 +1,44 @@
+
+    // Добавляем связь ментор-менти
+    @PostMapping("/{mentorId}/add/{menteeId}")
+    public void addMentorship(@PathVariable long mentorId, @PathVariable long menteeId) {
```
Лучше во всех методах контроллера написать вместо void - ResponseEntity<Void>. Так будет лучше, чтобы на той стороне клиенту возвращался какой-то статус этой операции (200, 400, 500). Естественно надо будет добавить return, но вот тут я не совсем помню что надо написать. Я думаю у гпт спросишь и он тебе чётче объяснит мой тейк. В принципе можно оставить и так, на вариант с возвратом статуса будет корректнее и удобнее фронтэндерам на той стороне

vzeten replied:
Супер, вот это подгон полезной инфы, вообще не былл в курсе что можно void так обернуть. Помог 🤝

Kalex-k commented on src/main/java/school/faang/user_service/controller/mentorship/MentorshipController.java:
```diff
@@ -0,0 +1,44 @@
+public class MentorshipController {
+    private final MentorshipService mentorshipService;
+
+    // Добавляем связь ментор-менти
```
Старайся избегать комментариев, названия методов и переменных должны уже быть достаточно понятными для понимания. Комментарии в коде это дурной тон, если это не договоренность команды.

vzeten replied:
Спасибо!

Commit:
добавил javadoc в интерфейс MentorshipService
добавил логи через аннотацию ломбок @Slf4j в MentorshipServiceImpl
переместил метод authorizeAndLoadUsers вниз под публичные методы в MentorshipServiceImpl
поправил в классе MentorshipController наименование параметров в аннотациях @GetMapping c UserId на userId

Kalex-k commented on src/main/java/school/faang/user_service/service/mentorship/MentorshipServiceImpl.java:
```diff
@@ -0,0 +1,85 @@
+        long currentUserId = userContext.getUserId();
+
+        if (currentUserId != mentorId && currentUserId != menteeId) {
+            throw new ForbiddenException("Доступ запрещен");
```
я бы писала бы сообщение подробнее с указанием причины отказа в доступе, что бы не гадать в дальнейшем

Kalex-k commented on src/main/java/school/faang/user_service/service/mentorship/MentorshipServiceImpl.java:
```diff
@@ -0,0 +1,85 @@
+
+    @Override
+    public void addMentorship(long mentorId, long menteeId) {
+        MentorshipPair pair = authorizeAndLoadUsers(mentorId, menteeId);
```
В задании сказано, что лучше писать логи в сервисах, можно добавить для наглядности  log.info("text");

Kalex-k commented on src/main/java/school/faang/user_service/service/mentorship/MentorshipServiceImpl.java:
```diff
@@ -0,0 +1,85 @@
+    public void addMentorship(long mentorId, long menteeId) {
+        MentorshipPair pair = authorizeAndLoadUsers(mentorId, menteeId);
+        pair.mentee.getMentors().add(pair.mentor);
+        mentorshipRepository.save(pair.mentee);
```
Посмотри что у тебя возвращает твой репозиторий. В нем User Entity возвращает всегда List<User>, что не как не гарантирует тебе отсутствие дублирования, но даже если бы у тебя был бы тут Set и ты избежал бы дублирования, то add() у тебя ничего не вернет, операция произойдет с результатом false, но для пользователя это будет невидно. Добавь явную проверку на дублирование и с логом с предупреждением об ошибке.

Kalex-k commented on src/main/java/school/faang/user_service/service/mentorship/MentorshipServiceImpl.java:
```diff
@@ -0,0 +1,85 @@
+    }
+
+    @Override
+    public void deleteMentorship(long mentorId, long menteeId) {
```
можно тоже добавить лог

Kalex-k commented on src/main/java/school/faang/user_service/service/mentorship/MentorshipServiceImpl.java:
```diff
@@ -0,0 +1,85 @@
+        }
+
+        return mentors.stream()
+                .filter(Objects::nonNull)
```
зачем тебе фильтр, если у тебя уже есть проверка на null твоего листа

vzeten replied:
но ведь лист не равно его запись, может быть каким-то способом сюда может залететь null?

Commit:
добавил в javadoc пустую строку перед параметрами

7glebalyot commented on src/main/java/school/faang/user_service/service/mentorship/MentorshipService.java:
```diff
@@ -0,0 +1,18 @@
+
+import java.util.List;
+
+@Service
```
Вот тут аннотация @Service бессмысленна. Spring ее обойдет, так как интерфейс это не объект. Аннотации можно вешать на классы, методы и ещё что-то. @Service можно повесить например на класс, который имплементирует этот интерфейс, что у тебя вроде бы и сделано. Но вот интерфейс точно нет, так что смело убирай эту строку.

vzeten replied:
Удалил

7glebalyot commented on src/main/java/school/faang/user_service/service/mentorship/MentorshipServiceImpl.java:
```diff
@@ -0,0 +1,85 @@
+    private final MentorshipRepository mentorshipRepository;
+    private final UserMapper userMapper;
+
+    private MentorshipPair authorizeAndLoadUsers(long mentorId, long menteeId) {
```
У тебя этот метод "authorizeAndLoadUsers" делает сразу 3 вещи. Я бы это все таки разделил, на отдельные методы. Потом эти все три отдельных метода вызвал бы в методе "authorizeAndLoadUsers". Короче вырезать 3 метода из этого главного, написать их отдельно, и потом вызвать их все в уже пустом "authorizeAndLoadUsers"

7glebalyot commented on src/main/java/school/faang/user_service/service/mentorship/MentorshipServiceImpl.java:
```diff
@@ -0,0 +1,104 @@
+    private final UserMapper userMapper;
+
+    @Override
+    public void addMentorship(long mentorId, long menteeId) {
```
Добавь проверку на то что связь уже существует. А то если связь уже есть, то в БД спокойно добавиться дупликат. Возможно для нас уже постарались и добавили защиту на дурака и в БД сделали UNIQUE, который ограничивает от дубликата. Но даже с UNIQUE может выдать ошибку при добавлении копии. Возможно лучше будет подождать ответ ментора или посоветоваться на встрече

7glebalyot commented on src/main/java/school/faang/user_service/service/mentorship/MentorshipServiceImpl.java:
```diff
@@ -0,0 +1,104 @@
+    }
+
+    @Override
+    public void deleteMentorship(long mentorId, long menteeId) {
```
Тут тоже не хватает проверки на то, что связь вообще есть. Если ее нет, то в принципе ничего не произойдет, но тогда и метод зря будет работать

7glebalyot reviewed (commented):
Проверил

Commit:
- Изменил тип возврата контроллера с void на ResponseEntity<Void>
- Добавил проверки на null для коллекций из БД
- Добавил проверку дубликатов в addMentorship()
- Добавил проверку результата удаления в deleteMentorship()
- Добавил @Transactional на методы изменения
- Удалил лишние комментарии в контроллере

AlexeyAntipin commented:
Пул реквесты стоит называть понятно, чтобы по названию можно было понять, что тут происходит, так как именно то, что написано в названии пул реквеста, пойдет в историю коммитов

AlexeyAntipin commented on src/main/java/school/faang/user_service/service/mentorship/MentorshipServiceImpl.java:
```diff
@@ -0,0 +1,140 @@
+
+        List<User> mentors = pair.mentee.getMentors();
+        if (mentors == null) {
+            mentors = new ArrayList<>();
```
Можно перенести инициализацию в саму сущность, тогда получится избавиться от лишнего if

AlexeyAntipin commented on src/main/java/school/faang/user_service/service/mentorship/MentorshipServiceImpl.java:
```diff
@@ -0,0 +1,140 @@
+    private MentorshipPair authorizeAndLoadUsers(long mentorId, long menteeId) {
+        long currentUserId = userContext.getUserId();
+
+        if (currentUserId != mentorId && currentUserId != menteeId) {
```
Эту валидацию можно вынести на уровень контроллера

AlexeyAntipin commented on src/main/java/school/faang/user_service/service/mentorship/MentorshipServiceImpl.java:
```diff
@@ -0,0 +1,140 @@
+        return new MentorshipPair(mentor, mentee);
+    }
+
+    private record MentorshipPair(User mentor, User mentee) {}
```
Не совсем понимаю, зачем здесь добавлять пару. Кажется, что в нужных методах можно просто получить ментора и менти и работать с ними дальше, так код станет проще

AlexeyAntipin commented on src/main/java/school/faang/user_service/service/mentorship/MentorshipServiceImpl.java:
```diff
@@ -0,0 +1,140 @@
+
+        if (mentees == null || mentees.isEmpty()) {
+            log.debug("No mentees found for userId={}", userId);
+            return List.of();
```
Лучше вместо List.of возвращать Collections.emptyList(). В таком случае ты будешь возвращать константу, а не создавать новый объект каждый раз

AlexeyAntipin commented on src/main/java/school/faang/user_service/service/mentorship/MentorshipServiceImpl.java:
```diff
@@ -0,0 +1,140 @@
+
+        if (mentors == null || mentors.isEmpty()) {
+            log.debug("No mentors found for userId={}", userId);
+            return List.of();
```
То же самое, лучше константу использовать

AlexeyAntipin commented on src/main/java/school/faang/user_service/service/mentorship/MentorshipServiceImpl.java:
```diff
@@ -0,0 +1,140 @@
+        log.debug("Fetching mentees for userId={}", userId);
+
+        User user = mentorshipRepository.getByIdOrThrow(userId);
+        List<User> mentees = user.getMentees();
```
Ты вызывал этот метод через постман? Кажется, что тут ошибка может выброситься

Commit:
- заменил List.of() на Collections.emptyList()
- Перенес валидацию пользователя для методов добавления и удаления связи в БД mentorship из MentorshipServiceImpl в MentorshipController
- Изменил код в методах deleteMentorship и addMentorship
- Перенес инициализацию списка менторов и менти в Users
